### ISSUE TITLE
Need help making the example work (Renamed from: "Some info about this project")

### ISSUE THREAD

**Issue by netcorefan1** [cvetomir-todorov/PolyMessage#3]
Hello,
I come up here after searching for a framework I can use in all of my projects and that will work in most situations (in order to maintain a compact base). I have read the description of PM and I like it, but I have some questions:
1. Any chance to see a NuGet package? Compiling everything myself is something I would like to avoid.
2. Is this a project that is going to be maintained?
3. As of now I need IPC (in particular shared memory), but I see that IPC is a work in progress. All the sample code I see can also be applied to IPC? Are there any limitations?
4. Is there a way to define classes and interfaces without all of those obtrusive ServiceContract/OperationContract attributes? Not sure on how they do, but [some frameworks](https://github.com/zyanfx/Zyan) has found a way to do that.
5. There are some [libraries](https://github.com/cloudtoid/interprocess) that are almost memory heap allocations free and this is something I would like to take advantage of. What you can say on that matter?

Many thanks

**Comment by cvetomir-todorov** [comment]
Hello,

Thank you for being interested in PM by submitting its first issue! I would try to answer your questions below.

1. I have planned to add automated infrastructure for building, running the tests and publishing the NuGet packages. I accept recommendation in terms of the tooling for this. The more integrated with Github the better. In terms of the actual packages I'd like to have a base package `PolyMessage`. I am considering adding separate packages for the client and the server, which would require refactoring though, so probably not now. I would have two sets of packages since I want users to depend only on what they actually use. Let me know what you think?
  - Transports - `PolyMessage.Transports.Tcp`, `PolyMessage.Transports.Ipc`
  - Formats - `PolyMessage.Formats.MessagePack`, `PolyMessage.Formats.NewtonsoftJson`, `PolyMessage.Formats.ProtobufNet`, `PolyMessage.Formats.Utf8Json`

2. With enough interest this project is something I'd like to move forward based on the feedback from the community. Additionally my initial thoughts of improvement are listed below. How do you think we can make PM more popular?
  - The idea is to provide application-development-ready communication capabilities which are missing in .NET, e.g. something that wraps up TCP and other primitive APIs and adds support for newer transports such as QUIC.
  - Write up some guide for contributions - e.g. transports, formats, secondary features. The core functionality can also be worked on but would require more supervision initially.
  - Automated testing is fine for now but improving the infrastructure as described above is required.
  - Improve fault tolerance, reliability, throughput so that it is more ready for production.
  - I am considering adding additional interaction styles, for example an entirely asynchronous bi-directional one-to-one communication.
  - At some point donating to the project would be beneficial since I do have a day-time job and a personal life, despite the partial lockdown.
3. A big selling point is that both transports and message formats can be switched centrally in the setup. If you initially have components which are deployed on a single machine you would use the IPC transport. If, later, as your application grows you need to move the components to other machines you should ideally just replace the `PolyTransport` instance instantiated. The transports are designed to be compatible in terms of API, only their specific settings are different (e.g. `NoDelay` for TCP) which is fine since they are easily configurable separately.
4. Everyone has an opinion but I would certainly not call 1 attribute here and there obtrusive. Declaring the type of the operation with `[PolyRequestResponse]` is 1 line. If you have other (conflicting!) interaction styles you would have to indicate somehow which one it is. There are other means to achieve it - e.g. making all methods part of the contract (as the framework you are mentioning) or implementing an imperative APIs to achieve the same effect (see ProtobufNet link below). This is in the list of features (currently a Trello project) but it is secondary given that it just improves an existing functionality. Also it is something a contributor can work on, if you are interested. Additionally PM can be changed so that `[PolyMessage]` attributes can be ommitted since the types are already inspected in the operation, so thank you for helping me realize it. The `[DataContract]` and `[DataMember]` are needed only by the frameworks used for the message formats. If you are using [ProtobufNet](https://github.com/protobuf-net/protobuf-net) you can skip them as described by the author [here](https://stackoverflow.com/a/12312756/608971). If attributes are required there isn't much you can do.
5. Currently resources are mainly allocated when the session is established. Each send/receive uses `ArrayPool<byte>` for the buffers. The 3rd party timer facility may be heap allocating a bit but the whole thing is going to be replaced in the future. Message formatters probably also have some heap allocations which is out of my control as well. Other than that I have to step through the code in order to make sure I didn't miss a small detail. All these nice features of the RPC come at a cost. If you need superb performance functionality - a custom approach with the framework you mention looks like a better idea. Also I would like to point you to [ZeroMQ's .NET native implementaion called NetMQ](https://github.com/zeromq/netmq) and [NanoMSG Next Generation P/Invoke-based .NET implementation](https://github.com/jeikabu/nng.NETCore) which should be more production-ready compared to all the RPC's mentioned here. I know for a fact that ZeroMQ has an IPC transport and you should just make sure it uses shared memory for the performance benefits. Also you have to figure out how to enforce ZeroMQ's frames to use `byte` array buffers for the serialized messages from a pool. For receiving [this](https://stackoverflow.com/q/42069711/608971) is what I've used based on the answer of the NetMQ's author.

**Comment by netcorefan1** [comment]
Hello and thanks for your detailed response.

1.  An automated infrastructure is a good choice. I don't have any experience on this, but I often see GitHub open source projects triggering builds through AppVeyor. You may want to take a look. [Vanara](https://github.com/dahall/Vanara) is a library that I recently added to my projects and works in this way. Although libraries should be loaded only when referenced in code, keeping them as separate packages is something that everyone would appreciate. It helps to keep the code cleaner and focus only on the needed features. I suppose should also simplify an extensible architecture. Your list is good although I would not lose anymore time in NewtonsoftJson because now is obsolete compared to the other alternatives. I would keep it only if does not requires too much work. For binary serialization I would consider the implementation of [BinaryPack](https://github.com/Sergio0694/BinaryPack). Benchmarks demonstrated to outperforms MessagePack. I would still keep MP as default binary serializer (since BP has some [downsides](https://github.com/Sergio0694/BinaryPack#are-there-some-downsides-with-this-approach) that for some users cannot be accepted).
Regarding pure Json, I switched to [SpanJson](https://github.com/Tornhoof/SpanJson) after having discovered that it is much better in terms of allocation and speed compared to Utf8Json (which is actually people still believe is the best one). I had an interesting [discussion](https://github.com/Tornhoof/SpanJson/issues/146) with the author in the past where you can find some interesting info on that matter. I am surprised this library is not yet popular like Utf8Json, but I believe it will be just a matter of time. The author is a wonderful person. I recommend you to do the switch and I guarantee you that you will forget Utf8Json which does not even seem under development anymore (last commit is 9 june 2018 and issues are unanswered).

2. The first step is making the project available on Nuget. If someone is lucky to enter into your github (which returns so many results when searching for IPC and similar) he will see that he will have to compile the project himself and will immediately close the page. This is something that no one want and I have been tempted to do the same, but your project was too nice and I felt at least it deserved to talk with the author. You can post testing builds here and me (along with other people that will join this thread) will help you with testing and suggestions on improvements. I spent nearly a full day in trying to find for a library that will allow me to do everything I need. What every developer want to avoid is discovering that in a future project a library is unable to perform certain tasks which are only possible by introducing another third party dependency. This inevitably introduces complexities, new dependencies, new learning curves etc. If you can achieve this goal, then I believe this is enough to make the library the most popular.
- application-development-ready communication capabilities is what should be the primary focus. I don't know QUIC, but I suppose is another emerging high performant technology (it's really hard to stay updated with everything).
- A guide would be a good thing (and I would say unavoidable if usage is complex), but I would keep this as the final step.
- I only tried testing a couple of times and I never had the need, but I know this is bad and I should correct this bad habit. I am not sure how much work can require this, but I would not keep this as a priority, The priority is the code.
- I believe that asynchronous bi-directional one-to-one communication is something that people expect (see [vs-streamjsonrpc](https://github.com/microsoft/vs-streamjsonrpc), I am currently testing the lib and below will tell you my initial thoughts)
- Donations are important and something that every developer who put so much efforts in projects fully deserve. I saw in some projects some donation feature integrated with Github (actually I don't remember where UPDATE: [here](https://github.com/EasyHook/EasyHook) you can get some nice ideas)

3. Excellent! I saw you got my point!

4. Here I introduce you [vs-streamjsonrpc](https://github.com/microsoft/vs-streamjsonrpc) which is the library I am currently testing. As of now seems to be the best option. Not surprising, it comes from Microsoft and expected to become part of the framework. It's easy to use (well, pretty easy) and works fine. It supports MessagePack, named pipes, web sockets and stream (I am not sure but this one should open doors to everything). I don't see Shared Memory support, but seems to be there is an unofficial connector (I opened an [issue](https://github.com/microsoft/vs-streamjsonrpc/issues/629) to ask more info on how to join the two pieces together). The library seems to have full asynchronous support (at least in testing so it seemed to me). I didn't had to decorate the class, but only now I realized that it was because I was not using MessagePack. This is not a problem since I know that all decoration are related to MessagePack. What made me a bit confused is when I saw those decorations mixed with poly ones, data contracts ones etc. Mine is nothing more than a sensation of loss returned from my eyes and I should only comment on this only after having tried PM. Your idea of making all methods part of the contract or implementing an imperative APIs is good, but I am talking of concerns that may not even exist once I try PM.

5. Reducing third party dependencies is something that we all appreciate and I suppose this also applies to the timer facility. If things can be done without it, it would be a great benefit. As far as I know, MessagePack allocations are very low and the ones from BinaryPack even lower. So, I believe this should be something to not worry about. I am only asking myself why [Interprocess](https://github.com/cloudtoid/interprocess) make the nearly zero [allocations](https://github.com/cloudtoid/interprocess#performance) as the main strength. I am afraid that we can only have some idea when it will be possible to do some benchmark. You're right, I forgot the RPC side and all its nice things are simply indispensable.
Thanks for the links. I have never taken seriously all those ZeroXX, NanoXX etc etc. I always believed they were only designed for Http or any other kind on communication outside the local network. Are you sure? I checked the links, but didn't found anything. Thanks also for the StackOverflow link, it's illuminating.
How you would compare those solutions with vs-streamjsonrpc? And, of course, with PolyMessage at its current status? At first it seemed to be a bit overkill (vs should stand for Visual Studio and I suppose this is why there are some VS dependencies, will need to check if there are other versions), but apart from this there is something that made me worried. I am not sure if this is related to what you was saying at the beginning of point 5:

https://user-images.githubusercontent.com/63934725/104390435-f95e3980-553d-11eb-81de-da1941b64fa6.mp4


As you can see, the connection itself is fast, but when I call for the first time the method "GetAllIngredientsAsync()", it takes one second to return the result. Subsequent calls return instantly. You're seeing release, in debug mode it's even worse, I have to wait a couple of seconds, the time required from VS to load the required modules. This happens with pipes and I am not sure if it will also applies with shared memory. For a longer session is a not a problem this initial load time, but when I need to launch a process, quickly execute a method and close, then it's something that I don't expect to happen especially betwen two local processes fully cooperating.
It looks like it is doing doing some sort of lazy loading, all the necessary stuff is loaded only when I execute the method for the first time. I am trying to see how I can force loading all the needed stuff from beginning of the communication, but I have yet to find a way. I am new to IPC, is this the way they are supposed to work? Or I am missing something obvious? Or it's just vs-streamjsonrpc that add additional overhead? Did you experience the same with PolyMessage?

Regarding the Zyan project (one of the ones that seemed to me the best projects along with yours), the author just [replied](https://github.com/zyanfx/Zyan) me. Unfortunately the project is heavily based on Remoting infrastructure and he don't think it's possible to port to the new .Net. I am going to invite him to join this thread because I see this as a chance of cooperating with you into this new amazing project.
Thanks

**Comment by netcorefan1** [comment]
I just want to post some updates. The slowness in vs-streamjsonrpc video seems due to "Anonymously Hosted DynamicMethods Assembly" loading. Even if I load all the references in advance, as soon as I call the first method, it starts the loading of this module. This is the cause of slowness and I am not sure what can be done.

I also wanted to try PolyMessage and unfortunately has not been a good beginning. I compiled the lib and despite I just used the same sample code published from you, I got a System.NullReferenceException. I didn't tried IPC because I wanted to first reproduce your sample code and then starting working from there.
The sample project: [PolyMessageTest.zip](https://github.com/cvetomir-todorov/PolyMessage/files/5816632/PolyMessageTest.zip)

I also wanted to experiment with the two libraries you recommended. NetMQ seems to support IPC only for Linux. There was a [discussion](https://github.com/zeromq/netmq/issues/331) about a possible implementation, but seems to be that they have not gone further. I couldn't make IPC working and I didn't spent further time on NetMQ and focused directly on NanoMsg (especially after having read comments on the web about its superiority). I must say that this library is extremely complicated. It has been very hard even to establish a basic connection. The client immediately returns a connection refused if the server has not yet been initialized. I suppose that NanoMsg should provide a mechanism to handle this, but I didn't found anything and the maximum I could do to quickly fix this was:
```
IReqSocket reqSock;
while (!factory.RequesterOpen().ThenDial(url).TryOk(out reqSock))
{
}
using (var s = reqSock)
{}
```

After lot of trials I have been able to deserialize a MessagePack message, but only by copying into an array:
`MessagePackSerializer.Deserialize<MyClass>(reply.AsSpan().ToArray())`
I checked your response in [SO](https://stackoverflow.com/questions/42069711/what-is-the-netmq-api-for-receiving-in-a-custom-buffer-in-order-to-avoid-creatin), but I have not been able to adapt your code into NanoMsg. There is a "RecvZeroCopy()" method which returns an IMemory interface which itself allows to call AsSpan<byte>, but the deserializer does not accept any Span<byte>. Any suggestion?

In the meantime, I got a response from the [vs-streamjsonrpc](https://github.com/microsoft/vs-streamjsonrpc/issues/629) guys and seems to be that there would be no benefits in performance compared to pipes, but only downsides.
I must admit that I am confused on what could be the right road.  vs-streamjsonrpc was the easier one to work with, but I saw it has great overhead. NanoMsg was fast, but the [API](https://nng.nanomsg.org/) complexity is terrifying!
Let me know how I can get PolyMessage working, I am very curious to see how compare!

**Comment by cvetomir-todorov** [comment]
I saw your PolyMessage test code. Unfortuntely you didn't really use the sample code quite correctly. 
* Your "server" actually starts a `PolyClient`, not a `PolyHost`.
* I think you didn't copy the actual **server** code.
* You should use strictly the `dev` branch, not the others ones that are in progress.
* To make things clear you should create a `Contracts` project where the messages and contract interface reside.
* You should move the service implementation to the server, not within the `Definitions.cs` file since it is an actual implementation.

I confirmed just now that when done correctly, as in the example, the request is sent and a response is received from the server. I would recommend to you to start the server in a separate console terminal manually rather than using `Process.Start`.

If you want to debug the client make sure you've increased the timeouts **on the server** so that idle clients are not disconnected after 30s, which is the default. And slow-receiving clients also have a timeout which I've set to infinite for your test:

```C#
PolyFormat format = new ProtobufNetFormat();
PolyTransport transport = new TcpTransport(new Uri("tcp://127.0.0.1:10678/"), loggerFactory);
transport.HostTimeouts.ClientSend = PolyTransport.InfiniteTimeout;
transport.HostTimeouts.ClientReceive = PolyTransport.InfiniteTimeout;

using PolyHost host = new PolyHost(transport, format, serviceProvider);
host.AddContract<IProductServiceContract>();

await host.StartAsync();
```

**Comment by cvetomir-todorov** [comment]
I want to let you know that I've decided to use Github Issues of my repository for **problems** with PolyMessage, just like your report that the example doesnt' work (which wasn't a real problem but whatever).

Github now offers [Discussions tab](https://github.com/cvetomir-todorov/PolyMessage/discussions) where a lot of this conversation could be moved to. Feel free to:
* Ask specific question in the Q&A section - e.g. is the project going to be maintained etc.
* Suggest an idea in the Ideas section - e.g. usage of other formats, transports, contracts. But make a separate conversation for each.
* For other stuff the General section is a fallback option

But I would leave this issue open until you've confirmed that the example works for you.

**Comment by netcorefan1** [comment]
Oh my God! I don't know where my head was when I wrote the server code! In these days I am doing so many tests with different libs and I need to decide for a definitive solution before these kind of mistakes multiplies!
Your sample now works well although the data exchange seemed slow, but may be is better to talk about this in the discussion tab.
I didn't applied the other changes you suggested (I wanted to first get it to work), so may be there is room for optimizations. I believe to have compiled the Dev branch because is the default one that appears in the browser. Sample used TCP transport (only now I realized that IPC transport support is on a different branch).
If you want you can close this issue, the problem has been solved.
